src/libxtp/openmp_cuda.cc

```diff
@@ -68,8 +63,26 @@ void OpenMP_CUDA::setOperators(const std::vector<Eigen::MatrixXd>&,
 }
 #endif
 
-bool isInVector(Index element, const std::vector<Index>& vec) {
-  return (std::find(vec.begin(), vec.end(), element) != vec.end());
+#ifdef USE_CUDA
+
+bool OpenMP_CUDA::isInVector(Index Id, const std::vector<GPU_data>& vec) {
+  return (std::find_if(vec.begin(), vec.end(), [&Id](const GPU_data& d) {
+            return d.Id == Id;
+          }) != vec.end());
+}
+bool OpenMP_CUDA::isGPUthread(Index ParentThreadId) const {
+  return isInVector(ParentThreadId, gpus_);
+}
+#endif
+
+Index OpenMP_CUDA::getParentThreadId() const {
+  return inside_Parallel_region_ ? threadID_parent_ : OPENMP::getThreadId();
+}
+Index OpenMP_CUDA::getLocalThreadId(Index ParentThreadId) const {
+  return inside_Parallel_region_ ? 0 : ParentThreadId;
+}
+Index OpenMP_CUDA::getNumberThreads() const {
+  return inside_Parallel_region_ ? 1 : OPENMP::getMaxThreads();
```

Comment:
This new functionality to manage the OpenMP/GPU threads improves significantly the code's readability. I am just wondering if in some point there should be some graph/schema explaining visually how the different parts interact. The code is getting more robust but there are significant more abstraction that is mostly in your head. The code quality is pretty good, I am just saying that maybe a graph/picture/schema may help to get the overall picture.

Reply:
Good point, this class will get uglier again unfortunately.